traversal/optimiser/Optimiser.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (C) 2021 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.traversal.optimiser;
+
+import com.google.ortools.linearsolver.MPSolver;
+import com.google.ortools.linearsolver.MPSolverParameters;
+import com.google.ortools.linearsolver.MPVariable;
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.ortools.linearsolver.MPSolverParameters.IncrementalityValues.INCREMENTALITY_ON;
+import static com.google.ortools.linearsolver.MPSolverParameters.IntegerParam.INCREMENTALITY;
+import static com.google.ortools.linearsolver.MPSolverParameters.IntegerParam.PRESOLVE;
+import static com.google.ortools.linearsolver.MPSolverParameters.PresolveValues.PRESOLVE_ON;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class Optimiser {
```

Comment:
wrapper around `MPSolver`:
records all the objective coefficients, variables, and constraints created and lazily instantiates them when calling `activate()`